### PR TITLE
EthernetNetworkModule for HPE OneView

### PR DESCRIPTION
**Pull request by fgbulsoni**
##### SUMMARY
HPE OneView is an infrastructure automation engine built with software intelligence. It streamlines provisioning and lifecycle management across compute, storage and fabric and enables IT staff to control resources programmatically through a unified API.

This PR aims to add the module for managing the [HPE OneView Ethernet Network resource](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html#rest/ethernet-networks): `EthernetNetworkModule` and its unit tests.

This resource has the `present` and `absent` states which will manage all of its aspects, including creating and updating the resources (using `present`) and removal of resources (using `absent`), and also features the `default_bandwidth_reset` state to reset a network bandwidth.

Related issue for more information: [HPE OneView support](https://github.com/ansible/ansible/issues/28354)

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`EthernetNetworkModule`

##### ANSIBLE VERSION
```
ansible 2.4.0 (hpe-oneview/ethernet-network 751b51553f) last updated 2017/08/17 13:56:59 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vagrant/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/dev/core_ansible_related/ansible/lib/ansible
  executable location = /home/vagrant/dev/core_ansible_related/ansible/bin/ansible
  python version = 2.7.9 (default, Aug  1 2017, 20:38:26) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION
Example of usage:
```yaml
- name: Ensure that the Ethernet Network is present using the default configuration
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      name: 'Test Ethernet Network'
      vlanId: '201'

- name: Update the Ethernet Network changing bandwidth and purpose
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      name: 'Test Ethernet Network'
      purpose: Management
      bandwidth:
          maximumBandwidth: 3000
          typicalBandwidth: 2000
  delegate_to: localhost

- name: Ensure that the Ethernet Network is present with name 'Renamed Ethernet Network'
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      name: 'Test Ethernet Network'
      newName: 'Renamed Ethernet Network'

- name: Ensure that the Ethernet Network is absent
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: absent
    data:
      name: 'New Ethernet Network'

- name: Create Ethernet networks in bulk
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: present
    data:
      vlanIdRange: '1-10,15,17'
      purpose: General
      namePrefix: TestNetwork
      smartLink: false
      privateNetwork: false
      bandwidth:
        maximumBandwidth: 10000
        typicalBandwidth: 2000

- name: Reset to the default network connection template
  oneview_ethernet_network:
    config: "{{ config_file_path }}"
    state: default_bandwidth_reset
    data:
      name: 'Test Ethernet Network'
```

### CODE REVIEW

**Comment by fgbulsoni** [comment]
I think the module is ready for review. Leaving my shipit here.

**Comment by fgbulsoni** [comment]
Updated PR after #28359 was merged, since that altered the path from `modules/remote_management/hpe` to `modules/remote_management/oneview`.
This is ready for review from someone from the core team, could someone have a look?
Thanks,

File: lib/ansible/module_utils/oneview.py
Comment (dagwieers) [inline lib/ansible/module_utils/oneview.py:149]: This seems weird. I thought we had just established data was a string ?

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: No trailing dot in short_description. Keep the index clean and tidy ;-)

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: The quotes here are unneeded and unwanted actually.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Make this a list, and drop the quotes :-)

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: No need to quote strings in YAML. (unless it's yes, no or null I guess)

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Make this:
```python
        argument_spec = dict(
            state=dict(type='str', required=True, choices=['absent', 'default_bandwidth_reset', 'present']),
            data=dict(type='dict', required=True),
        )
```
It's easier to read if there is one parameter per line, and we have 160 allowed columns on a line (and very wide screens these days).

I also sorted the options in the list (order is not important) and added the type as the first option (as this one should always be set anyway). So it looks nicer (to the eye of the beholder).

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Does it matter here ? I would expect to always set changed to True, and maybe append the additional message. Otherwise this gets lost, no ?

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: The interface of __present and __bulk_present work differently. One is passing the result dict, the other is setting it as a global variable. I think you should stick to one way of working to make it easier to understand what's going on.

Comment (dagwieers) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Actually result could be a property of the main object in this case.

**Review by dagwieers — CHANGES_REQUESTED** [review]
Not everything I suggest is required. Some general notes.
I'll re-review when you are ready.

File: lib/ansible/module_utils/oneview.py
Comment (fgbulsoni) [reply lib/ansible/module_utils/oneview.py:149]: `data` is an abstract field of type dict which will contain str, integers, lists and other dicts, which are returned or passed to HPE OneView.

Comment (dagwieers) [reply lib/ansible/module_utils/oneview.py:149]: Right, I see I have misread the if-else statements there. Sorry.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Duh, I keep on putting that dot. Will try to remember.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Good catch! 
I've removed the `if not result['changed']:` from here but left the update to `changed` and `msg`, since `changed` may eq to `False` here if nothing other than the `bandwidth` of the resource changed.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Modified __bulk_present to use `result`-like format now same as `__present`.

File: lib/ansible/module_utils/oneview.py
Comment (dagwieers) [inline lib/ansible/module_utils/oneview.py:155]: To be honest, I would leave it up to the Exception object to check if the 3 parameter is fine, rather than make 2 almost identical calls here. But that's just a mental note I made, not a requirement per se.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: So these examples don't have a `delegate_to: localhost`and I think you should have it there (unless you explain why/when it isn't needed).

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: So you put a variable there everywhere, but that's not really helpful. Most people would like to know what kind of file this is. What a typical naming-convention would be and where it would be typically located.

Should it be an absolute path, or if it's relative, relative to what ? Things you need to explain and demonstrate through these examples.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: So you added an extra maintainer. Very good :+1:

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: You could make list-items of these three lines. I think it comes out better in the documentation.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py]: Ops, missed this

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: Yep, didn't know `author` could be a list. We're the team that is maintaining the original [oneview-ansible](https://github.com/HewlettPackard/oneview-ansible), so it makes sense to add him as well.

Comment (dagwieers) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: The more the merrier. Take that literally, because if you have 2 maintainers adding 'shipit' to a PR, the PR is automatically merged. That means you need at least three maintainers if one of the maintainer creates a PR.

File: lib/ansible/module_utils/oneview.py
Comment (fgbulsoni) [reply lib/ansible/module_utils/oneview.py:155]: Changed this to leave up to the `Exception` and then changed back, it makes unit testing expectations a little bit uglier without the if. If this is acceptable as is, will leave it alone for now.

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: Interesting, adding the PO of the project so we reach the 3-person mark then. 
Thanks :octocat:

**Comment by fgbulsoni** [comment]
I *think* these last commits have covered everything that was requested in the review. @dagwieers , could you please verify?
Thanks,

File: lib/ansible/module_utils/oneview.py
Comment (dagwieers) [reply lib/ansible/module_utils/oneview.py:155]: ok

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: Hey, it is expected that the 2 maintainers doing the shipit are actually reviewing the code and are being a PITA during review just like me. If this is about doing favors, it's not going to help the cause :-)

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:29]: There is an indentation issue here. I don't know if the documentation is allowed to be a nested list...

**Review by dagwieers — APPROVED** [review]

File: lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py
Comment (fgbulsoni) [reply lib/ansible/modules/remote_management/oneview/oneview_ethernet_network.py:23]: Sure, the PO will be reviewing as well and we do weekly demos for functionality :octocat:  Just won't be actively committing code

**Pull request merged**
